forum/qa-plugin/q2a-googleauthenticator-login/q2a-googleauthenticator-overrides.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+function qa_set_logged_in_user($userId, $handle = '', $remember = false, $source = null)
+{
+    require_once QA_INCLUDE_DIR . 'app/cookies.php';
+    qa_start_session();
+    if (!isset($userId)) {
+        // logout
+        qa_report_event(
+            'u_logout',
+            qa_get_logged_in_userid(),
+            qa_get_logged_in_handle(),
+            qa_cookie_get()
+        );
+        qa_clear_session_cookie();
+        qa_clear_session_user();
+
+        return;
+    }
+
+    // login
+    require_once QA_INCLUDE_DIR . 'db/selects.php';
+    $result = qa_db_read_all_assoc(
+        qa_db_query_sub(
+            'SELECT 2fa_enabled FROM ^users WHERE userid = #', $userId
+        )
+    );
+    if (count($result) !== 1) {
+        echo 'Invalid num_rows';
+        die;
+    }
+    $usingTwoFactorAuth = (bool) $result[0]['2fa_enabled'];
+    if ($usingTwoFactorAuth && '2fa' !== $source) {
+        echo '
+<form method="post" id="form" action="./2fa-auth">
+	<input name="login" type="hidden" value="' . qa_post_text('emailhandle') . '">
+	<input name="password" type="hidden" value="' . qa_post_text('password') . '">
+	<input name="remember" type="hidden" value="' . qa_post_text('remember') . '">
+	<input name="redirect" type="hidden" value="' . $_GET['to'] . '">
```

Comment:
pobranie i podstawienie z GET bez żadnej walidacji i sprawdzenia czy istnieje?

Reply:
Cały czas szukam zasadności tego formualrza. Zastanowię się czy nie da się tego inaczej ugryźć.

Reply:
A czy zamiast formularza nie można posłać POSTa Ajaxem? Skoro ten formularz i tak jest automatycznie submitowany skryptem poniżej.

Reply:
Pewnie można. Na razie szedłem w stronę żeby to w ogóle uruchomić i uporządkować. Teraz trzeba to do ładu doprowadzić.